packages/edit-site/src/components/editor/global-styles-provider.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * External dependencies
+ */
+import { set, get } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext, useEffect } from '@wordpress/element';
+import { useEntityProp } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { getGlobalStyles } from './global-styles-renderer';
+
+const GlobalStylesContext = createContext( {
+	/* eslint-disable no-unused-vars */
+	getProperty: ( context, path ) => {},
+	setProperty: ( context, path, value ) => {},
+	globalContext: {},
+	/* eslint-enable no-unused-vars */
+} );
+
+export const useGlobalStylesContext = () => useContext( GlobalStylesContext );
+
+export default ( { children, baseStyles, contexts } ) => {
+	const {
+		userStyles,
+		getProperty,
+		setProperty,
+	} = useGlobalStylesFromEntities();
+
+	useGlobalStylesEffectToUpdateStylesheet( contexts, baseStyles, userStyles );
+
+	return (
+		<GlobalStylesContext.Provider
+			value={ {
+				getProperty,
+				setProperty,
+				contexts,
+			} }
+		>
+			{ children }
+		</GlobalStylesContext.Provider>
+	);
+};
+
+/**
+ * Hook that exposes an API to retrieve and update user styles.
+ *
+ * @param {number} entityId Entity that stores the user data as CPT.
+ *
+ * @return {Object} User data as well as getters and setters.
+ */
+const useGlobalStylesFromEntities = () => {
+	const [ content, setContent ] = useEntityProp(
+		'postType',
+		'wp_global_styles',
+		'content'
+	);
+	const userStyles = content ? JSON.parse( content ) : {};
+
+	const getProperty = ( context, path ) =>
+		get( userStyles?.[ context ]?.styles, path );
+
+	const setProperty = ( context, newValues ) => {
```

Comment:
On each global style change e.g: each time we press arrow up to change font size, we will be serializing a JSON string. We have the concept of transient attributes on entities that fits this use case well. E.g: blocks are HTML strings but on each update to a block we don't serialize blocks we just update a structure and serialize on save. I guess we should use the same mechanisms here.

Reply:
I like that idea and actually tried to do it when this was using the entities API directly instead of `useEntity`. I'll give it a try. I was thinking this would be a good fit for a follow-up PR.

Reply:
Created a follow-up PR at https://github.com/WordPress/gutenberg/issues/25138